Crm457 1146/permissions

dwhenry opened the pull request:
## Description of change
Add permission to updating the application

## Link to relevant ticket

[CRM457-1145](https://dsdmoj.atlassian.net/browse/CRM457-1145)


[CRM457-1145]: https://dsdmoj.atlassian.net/browse/CRM457-1145?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
Append events instead of overwritting

This is to ensure that events are not deleted.

An issue with updates to the events JSON field from being saved was encountered.
Multiple attempts were made to resolve this issue including the use of
sqlalchemy.ext.mutable and options details https://stackoverflow.com/questions/60893464/nested-json-array-not-updating-in-db/60927486#60927486
however none of these proved successful. As such we simple mark the field
as dirty if additional events are inserted.

Commit:
Convert JSON fields to JSONB

Commit:
Use JSONB in the schema def files

Commit:
fix

Commit:
remove code related to notifier class

Commit:
Add permission layer based on user roles

Commit:
More complete states list

Commit:
update role names with capitalisation

Commit:
better comments

patrick-laa commented on laa_crime_application_store_app/services/permissions.py:
```diff
@@ -0,0 +1,39 @@
+
+
+class Permissions:
+    LOCKED_STATES = ("granted", "expired", "auto_grant")
```
I think CRM7 states `further_info` and `provider_requested` are also locked (or maybe they're provider-editable is we think the provider could in theory amend and re-submit them in the future)

patrick-laa commented on laa_crime_application_store_app/services/permissions.py:
```diff
@@ -0,0 +1,39 @@
+class Permissions:
+    LOCKED_STATES = ("granted", "expired", "auto_grant")
+    CASEWORKER_EDITABLE_STATES = ("submitted", "provider_updated")
+    PROVIDER_EDITABLE_STATES = ("part_granted", "rejected", "part_grant", "sent_back")
```
We've killed off `part_granted` now. It shouldn't exist anywhere.

patrick-laa commented on laa_crime_application_store_app/services/permissions.py:
```diff
@@ -0,0 +1,39 @@
+
+class Permissions:
+    LOCKED_STATES = ("granted", "expired", "auto_grant")
+    CASEWORKER_EDITABLE_STATES = ("submitted", "provider_updated")
```
This needs to include `sent_back` because the caseworker app is responsible for updating `sent_back` applications to `expired` via a daily scheduled job.

dwhenry replied:
so this break out rule of only one app can edit in each state... :(

patrick-laa commented on laa_crime_application_store_app/services/v1/application_service.py:
```diff
@@ -123,6 +125,9 @@ def update_existing_application(
         ):
             return existing_application.id
 
+        if not get_permissions().allow_update(existing_application, roles):
```
We should have some sort of locking here. At the moment, it's not impossible that the provider app would try to update a `sent_back` PA application at the precise moment the caseworker app would try to mark it as expired. We don't want them both to pass the permission check simultaneously and then both apply their updates to the same record.

dwhenry replied:
So not going to worry about this in this PR as due to one being an automated task and the other manual submission there is low likelihood of overlap with submission times - if we are worries I would suggest changing the expiry task from 11pm to 4am.

patrick-laa replied:
Is it not like a 2-line addition to add a 'SELECT FOR UPDATE'? The point of these PRs is to protect data integrity, and I'd rather our strategy not be "hope a lawyer doesn't pull an all-nighter"

Commit:
filter by ID in tests to ensure correct object is returned

Commit:
Use Depends, refactor

Commit:
Move permissions settings into a YAML file

Commit:
sync issues

Commit:
Use with_for_update to enable locking

jogilder commented on laa_crime_application_store_app/services/permissions.py:
```diff
@@ -0,0 +1,47 @@
+
+
+def validate_can_create(user: User = Depends(azure_schema)) -> None:
+    if user is None or not user.roles:
```
Didn't we say that we would have a check for whether or not the authentication was set here?
`settings.authentication_required.lower() != "true"`

jogilder replied:
also would be better to not have this as an `elif`, can we just return from this if block and then have the next block as another if?

jogilder replied:
And we won't remove this completely when the roles are implemented as we will still need to check if we have authentication turned on or not.

jogilder commented on laa_crime_application_store_app/services/permissions.py:
```diff
@@ -0,0 +1,47 @@
+) -> None:
+    if application.application_state in permissins.locked:
+        raise InvalidAuth("Application in locked state")
+    elif not user.roles:
```
won't this fail if you have authentication turned off?

jogilder replied:
Also if you wanted to we can use a match statement here as allowed from python 3.10 :)
https://docs.python.org/3/whatsnew/3.10.html#pep-634-structural-pattern-matching

jogilder commented on laa_crime_application_store_app/services/permissions.py:
```diff
@@ -0,0 +1,47 @@
+def validate_can_update(
+    application: Application, user: User = Depends(azure_schema)
+) -> None:
+    if application.application_state in permissins.locked:
```
typo in permissins

jogilder commented on laa_crime_application_store_app/config/permission_settings.py:
```diff
@@ -0,0 +1,24 @@
+    model_config = SettingsConfigDict(yaml_file=f"{THIS_DIR}/permissions.yaml")
+
+    locked: List[str]
+    casework_editable: List[str]
```
```suggestion
    caseworker_editable: List[str]
```

jogilder commented on laa_crime_application_store_app/config/permission_settings.py:
```diff
@@ -0,0 +1,24 @@
+    casework_editable: List[str]
+    provider_editable: List[str]
+    all_editable: List[str]
+    casework_role: str
```
```suggestion
    caseworker_role: str
```

jogilder commented on laa_crime_application_store_app/services/permissions.py:
```diff
@@ -0,0 +1,47 @@
+from laa_crime_application_store_app.models.application_schema import Application
+from laa_crime_application_store_app.services.auth_service import azure_schema
+
+permissins = get_permissions()
```
```suggestion
permissions = get_permissions()
```

jogilder commented on laa_crime_application_store_app/config/permissions.yaml:
```diff
@@ -0,0 +1,16 @@
+locked:
```
Have we got all states here?

jogilder commented on laa_crime_application_store_app/internal/notifier.py:
can you document why this is removed in the conversation? I assume it is not used.

jogilder commented on laa_crime_application_store_app/services/v1/application_service.py:
```diff
@@ -168,7 +177,12 @@ def update_existing_application(
+        return (
+            db.query(Application)
+            .filter(Application.id == app_id)
+            .with_for_update()
```
can you explain that this is added for row locking? And what error do we get if someone tries to edit at the same time and how is it captured and dealt with?

jogilder replied:
There should be a test for race condition...

jogilder commented on tests/routers/v1/application_test.py:
```diff
@@ -88,11 +87,7 @@ def test_data_selected_version_returns_400(client: TestClient, seed_application)
-    mock_notify, client: TestClient, dbsession: Session
-):
-    mock_notify.return_value = True
+def test_post_application_returns_200(client: TestClient, dbsession: Session):
```
this is tricky because to my mind this test should fail as you have no roles (roles is an empty list) - The test should pass with a provider role and fail with caseworker or anything else as here we are mocking what happens when there is an authenticated user.
The issue is that we have methods that allow us to continue if there are no roles.

jogilder replied:
We may have to feature flag this.

jogilder commented on tests/services/conftest.py:
```diff
@@ -0,0 +1,29 @@
+from fastapi_azure_auth.user import User
+
+
+def build_user():
```
would be good to name this something different to that in the other conftest but not essential.

jogilder commented on tests/services/permissions_test.py:
```diff
@@ -0,0 +1,109 @@
+)
+
+
+def test_can_create_when_no_roles(normal_user):
```
I don't think this test should be in here and maybe we feature flag this work instead.

jogilder commented on tests/services/permissions_test.py:
```diff
@@ -0,0 +1,109 @@
+    assert validate_can_create(normal_user) is None
+
+
+def test_can_update_when_no_roles(normal_user):
```
ditto this test

jogilder reviewed (commented):
@dwhenry sorry lots of comments.

dwhenry committed:
Permissions spike (#111)

## Description of change

## Link to relevant ticket

[CRM457-](https://dsdmoj.atlassian.net/browse/CRM457-)

## Notes for reviewer

## Screenshots of changes (if applicable)

### Before changes:

### After changes:

## How to manually test the feature

---------

Co-authored-by: Jo Gilder <3441835+jogilder@users.noreply.github.com>

Commit:
attempt at working tests

Commit:
Stuff

Commit:
idea failure

Commit:
mostly working

Commit:
request stuff not working

jogilder committed:
post tests working need to do more on update as think we need to pass in roles to tests

jogilder committed:
remove brackets for depends

patrick-laa commented:
Closing as we've done this in Ruby instead
